modules/router-store/src/router_store_module.ts

```diff
@@ -141,12 +206,28 @@ export function _createDefaultRouterConfig(
 
   return {
     stateKey: DEFAULT_ROUTER_FEATURENAME,
+    serializer: DefaultRouterStateSerializer,
+    dispatchNavActionOnEnd: false,
     ..._config,
   };
 }
 
+export function _createSerializer(
+  config: StoreRouterConfig
+): RouterStateSerializer {
+  // This function gets handed a complete config-object from _createRouterConfig,
+  // so we know the serializer property exists
+  return new config.serializer!();
```

Comment:
This doesn't support dependencies being injected into the serializer

Reply:
Is this even possible without jumping through fire hoops? The problem is that we don't know the dependencies, so we somehow have to get them dynamically. But to my knowledge, getting these and making sure they exist (and those who are transitively required, too) is not really supported by Angular through some public API. It is only supported to statically define the dependencies.

So we would end up either using Angular internals which is brittle, or write the DI-mechanisms again.

If this is not really possible, maybe we should ditch the serializer option..

@timdeschryver any ideas?

Reply:
I didn't have the time to take this all in, I'll try to get back at this later today.

Reply:
It's already being injected in the constructor, which will resolve its dependencies. Just use the config to tell it which one to provide. Remove the `StoreRouterConfigFunction` and move all the providers into the `forRoot` method. It will be breaking, but more consistent

```
  static forRoot(
    config: StoreRouterConfig | StoreRouterConfigFunction = {}
  ): ModuleWithProviders {
    return {
      ngModule: StoreRouterConnectingModule,
      providers: [
          { provide: _ROUTER_CONFIG, useValue: config },
          { provide: RouterStateSerializer, useClass: config.serializer ? config.serializer : DefaultRouterStateSerializer }
     ],
    };
  }
```

Reply:
Ok when there is no function to execute it will work (with AOT, too).